### PR TITLE
CMake ExternalProject: Fix unable to find 'hiredis' headers even when 'find_package' found it 

### PR DESCRIPTION
**Pull request by twoixter**
The problem started after v1.2.2, when #188 was introduced. Importing redis++ as a ExternalProject no longer worked with `hiredis` installed from macPorts (macOS).

### Environment information

- **macOS 10.15.7** (Catalina) A bit old, but irrelevant for this issue
- **CMake 3.21.4**
- **Apple clang version 12.0.0** (clang-1200.0.32.29)
- **macPorts version 2.7.1**
- **hiredis version 1.0.2**

### Steps to reproduce
- Install `hiredis` with `sudo port install hiredis`. This will install `hiredis` in `/opt/local/include/hiredis`.
- All additional tools should be installed like cmake, clang, and all that stuff.

Use the following CMakeList.txt (just as example, from memory, just fill in the gaps):

```
include(ExternalProject)
set(REDIS_PLUS_PLUS_INSTALL_LOCATION ${CMAKE_BINARY_DIR}/redis_plus_plus)
ExternalProject_Add(redis_plus_plus
    GIT_REPOSITORY https://github.com/sewenew/redis-plus-plus.git
    GIT_TAG 1.3.0
    GIT_SHALLOW TRUE
    CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${REDIS_PLUS_PLUS_INSTALL_LOCATION}
    CMAKE_ARGS -DCMAKE_INSTALL_MESSAGE=LAZY
    CMAKE_ARGS -DREDIS_PLUS_PLUS_CXX_STANDARD=14
    CMAKE_ARGS -DREDIS_PLUS_PLUS_BUILD_SHARED=OFF
    CMAKE_ARGS -DREDIS_PLUS_PLUS_BUILD_TEST=OFF
)
```

Trying to compile a project using the above ExternalProject yields the following error later after `make`:

```
~/build2/redis-test/redis_plus_plus-prefix/src/redis_plus_plus/src/sw/redis++/connection.h:27:10: fatal error:
      'hiredis/hiredis.h' file not found
#include <hiredis/hiredis.h>
         ^~~~~~~~~~~~~~~~~~~
```

I've found that although the original CMakeList finds the `hiredis` package with `find_package`, then the include directory is not added with `target_include_directories`.

I've also first tried several combinations of `-DCMAKE_PREFIX_PATH` as you suggest in the README, but to no avail. The only solution that worked was adding the `${hiredis_INCLUDE_DIRS}` at the right place.

### CODE REVIEW

**Comment by sewenew** [comment]
@twoixter It seems that you should also add `CMAKE_ARGS -DCMAKE_PREFIX_PATH=/opt/local` to `ExternalProject_Add`.

Take the following CMakeLists.txt for example:

```
cmake_minimum_required(VERSION 3.1)

project(app)

set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")

include(ExternalProject)
set(REDIS_PLUS_PLUS_INSTALL_LOCATION ${CMAKE_BINARY_DIR}/redis_plus_plus)
ExternalProject_Add(redis_plus_plus_download
    GIT_REPOSITORY https://github.com/sewenew/redis-plus-plus.git
    GIT_TAG 1.3.0
    GIT_SHALLOW TRUE
    CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${REDIS_PLUS_PLUS_INSTALL_LOCATION}
    CMAKE_ARGS -DCMAKE_PREFIX_PATH=/opt/local                ##### <------- specify hiredis installation path
    CMAKE_ARGS -DCMAKE_INSTALL_MESSAGE=LAZY
    CMAKE_ARGS -DREDIS_PLUS_PLUS_CXX_STANDARD=14
    CMAKE_ARGS -DREDIS_PLUS_PLUS_BUILD_SHARED=OFF
    CMAKE_ARGS -DREDIS_PLUS_PLUS_BUILD_TEST=OFF
)
SET(REDIS_PLUS_PLUS_INCLUDE ${REDIS_PLUS_PLUS_INSTALL_LOCATION}/include)
SET(REDIS_PLUS_PLUS_LIB ${REDIS_PLUS_PLUS_INSTALL_LOCATION}/lib)

add_library(redis_plus_plus SHARED IMPORTED)
set_target_properties(redis_plus_plus PROPERTIES IMPORTED_LOCATION ${REDIS_PLUS_PLUS_LIB}/libredis++.a)

add_executable(app a.cpp)

find_package(hiredis)
if(hiredis_FOUND)
        target_link_libraries(app PRIVATE hiredis::hiredis)
else()
        # in case hiredis is not installed with hiredis
        find_path(HIREDIS_INCLUDE hiredis)
        find_library(HIREDIS_LIB hiredis)
        target_include_directories(app PRIVATE ${HIREDIS_INCLUDE})
        target_link_libraries(app PRIVATE ${HIREDIS_LIB})
endif()

target_include_directories(app PRIVATE ${REDIS_PLUS_PLUS_INCLUDE})
target_link_libraries(app PRIVATE redis_plus_plus)

add_dependencies(app redis_plus_plus_download)
```

Since hiredis is installed at /opt/local, you should run cmake the following command: `cmake -DCMAKE_PREFIX_PATH=/opt/local ..`.

Regards

**Comment by twoixter** [comment]
Hi @sewenew !! Sorry to reply so late... I somewhat missed your reply notification. I've just tested with recent changes on your repository to see if my changes still apply. And they do, it's still failing. Ok, here is the thing:

The directory `/opt/local` is standard in macOS with MacPorts, so the installed `cmake` already takes this directory into account as a system directory. That's why adding `CMAKE_ARGS -DCMAKE_PREFIX_PATH=/opt/local` is somewhat redundant. It is much like adding `CMAKE_ARGS -DCMAKE_PREFIX_PATH=/usr/include`.

Anyway, I did gone ahead and tried your suggestion with no avail. This is expected, as `/opt/local` is a standard directory with MacPorts' Cmake.

Another problem with your recommendation is that this CMakeFile is also compiled in Linux and under CI/CD (also Linux). So 
we would be forced to use `if (APPLE)` or similar to branch the `ExternalProject_Add`. With my suggested changes, it makes your CMakeLists.txt truly system independent, even in macOS whenever you use MacPorts or Homebrew.

For us it is kind of a regression, since it was working without problems before v1.2.2. Will it work for you if I change instead of using `... PUBLIC ${hiredis_INCLUDE_DIRS}` the code reads `... PUBLIC $<BUILD_INTERFACE:${HIREDIS_HEADER}>` ?

**Comment by sewenew** [comment]
@twoixter Does your hiredis version is older than v1.0.0?

I'm not an expert of CMake, so I did some research these days. Looks like that there're two forms of `find_package` (Check [this](https://stackoverflow.com/questions/35815850/getting-imported-targets-through-find-package) for detail). Since hiredis v1.0.0, it supports config-file packages, i.e. [commit ec08c2b94a](https://github.com/redis/hiredis/commit/ec08c2b94a6d2e3f3572dfdac734fb8a92682ac8#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20a). With this feature, you don't need to specify `target_include_directories` explicitly.

redis-plus-plus's CMakeLists.txt is based on this feature, and I was not aware that this feature was introduced recently. So if you installed an old version of hiredis, you might get the problem.

I'm not sure if there's any side effect if we add `target_include_directories` when user installs the latest version of hiredis (which supports config-file package). I'll do some more research to ensure the CMakeLists.txt be compatible with old and new version of hiredis. If you have any idea on it, feel free to let me know.

Regards

**Comment by twoixter** [comment]
Hi @sewenew !! Thanks a lot for your comments. Knowing that `hiredis` recently changed to use config-file packages I also did my own research. Since we mix macOS with Ubuntu (mainly 20.04) into our workflow, I found some interesting insights and also learned quite a bit.

Ubuntu 20.04 is quite conservative about package versions, for good reasons. The latest `hiredis` version available in its apt is 0.14, and the latest `cmake` available is 3.16. Both of which worked. This is no surprise since 0.14 still used `pkg-config` and thus activated the `else(hiredis_FOUND)` in your code later triggering the `target_include_directories`. So far, so good.

Macports is not so quite conservative, so in macOS versions are the latest available at the moment: 1.0.2 for `hiredis` and 3.22.1 for `cmake`. This is were the thing got interesting... :-) I manually installed both latest `cmake` and `hiredis` in Ubuntu for the sake of comparison and... also worked (this time using the `find_package` feature). So far, but not so good since this should match the macOS configuration.

After many hours fiddling with `g++` and `clang` to see if different compilers made different outcomes, I finally got to the real problem. And here comes the **TL;DR**: 

I saw `find_package` work both in macOS and Linux, using both standard `/usr/local` and non-standard `/opt/local` directories. For this to work, the path `/opt/local` must be included in the `CMAKE_SYSTEM_PREFIX_PATH` variable. For instance, macOS `cmake` have a very long and complicated `CMAKE_SYSTEM_PREFIX_PATH`.

CMake then manages to automatically include `hiredis_INCLUDE_DIRS` in the target include directories. What I saw, and this is the **only difference** between Linux and macOS, is the use of the `-isystem` flag to the compiler. The Linux `cmake` correctly pass this flag to the compiler as `-isystem /opt/local/include`, **but** Macports `cmake` pass this flag as `-isystem /opt/local/.`. I don't know why the Macport version uses `/opt/local/.` instead of `/opt/local/include`.

On the other hand, adding `target_include_directories(...)` as suggested in my PR did not make any difference in Linux but made the macOS compilation work. I checked this and the compilation command line argument did not change adding `target_include_directories` in Linux. I suspect this is why: `target_include_directories` adds new search directories to targets but I think `cmake` collapses then so it does not add a new directory if it is already present.

In other words, if the `/usr/local/include` directory is already present by using `find_package`, then adding it again with `target_include_directories` have no effect. However, since it is `/opt/local/.` in macOS from `find_package`, I get the correct `-I/opt/local/include` making the full command line as `-I/opt/local/include -isystem /opt/local/.`.

So, next steps. I see my PR is redundant as long as the problem lies really with Macports version of `cmake`, it is not a problem on older `hiredis` or something. On the other hand, it does no harm to add the proposed `target_include_directories`. I will be happy to close the PR if you like, and I'll try to check why my `cmake` behaves like that.

**Comment by sewenew** [comment]
@twoixter Thanks a lot for your detailed research! I also learned quite a bit from your comment :)

It looks like that it's a problem of Macport version of cmake. I'd like to keep this PR open, and in case others might meet the same problem or have some new idea on it. If you make any progress on it, please feel free to let me know.

Thanks again!

Regards

**Comment by sewenew** [comment]
@twoixter I did some research, and finally, I got the idea why we need to call `target_include_directories` even if `find_package` finds hiredis.

Because hiredis' CMakeLists.txt does not set `INCLUDES DESTINATION` when exporting hiredis-target:

```
INSTALL(TARGETS hiredis
    EXPORT hiredis-targets
    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})       # <----  `INCLUDES DESTINATION include`   is NOT set
```

That's why you got `-isystem /opt/local/.` flag instead of `-isystem /opt/local/include`. Check [this](https://cmake.org/cmake/help/latest/command/install.html#installing-targets) and [this](https://cmake.org/cmake/help/v3.19/guide/importing-exporting/index.html#exporting-targets) for more info.

I've merge your PR, and made some minor change, i.e. add `BUILD_INTERFACE`. The latest code has been merged into master branch. You can have a try.

Thanks for finding this problem!

Regards
